feat: TimerEventListener implementation first iteration

subhrajyotim opened the pull request:
Hello Roman,
The first iteration of the timer event listener.

Thanks,
Subhro.

romansmirnov commented:
Hi Subhro,

Thanks for your pull request. It looks great so far. I will dig into it next week and merge it!

In the meantime we could discuss the next step in the forum.

Thanks for your effort.

Cheers,
Roman

subhrajyotim commented:
Yes Roman, looking fwd to the next steps.

Thanks for all the guidance and patience.

Thanks,
Subhro.

On Fri, Feb 12, 2016 at 1:13 PM, Roman Smirnov <notifications@github.com>
wrote:

> Hi Subhro,
>
> Thanks for your pull request. It looks great so far. I will dig into it
> next week and merge it!
>
> In the meantime we could discuss the next step in the forum.
>
> Thanks for your effort.
>
> Cheers,
> Roman
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/camunda/camunda-bpm-platform/pull/201#issuecomment-183218275>
> .
>

romansmirnov commented on engine/src/main/java/org/camunda/bpm/engine/impl/cmmn/handler/TimerEventListenerItemHandler.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * @author subhro
+ */
+public class TimerEventListenerItemHandler extends ItemHandler{
```
Instead of extending the `ItemHandler` extend the already existing `EventListenerItemHandler`

romansmirnov commented:
Hi Subhro,

I am very sorry for the late response. I did not receive any notification from Github about your answer. Sorry!

I had a look into your pull request. There are some small adjustments necessary (see the comments I made). Could you adjust the pull request according the comments I made please? Furthermore, please change our coding style (see https://github.com/camunda/camunda-bpm-platform/tree/master/settings/eclipse).

The next step would be to create a [JobDeclaration](https://github.com/camunda/camunda-bpm-platform/blob/master/engine/src/main/java/org/camunda/bpm/engine/impl/jobexecutor/JobDeclaration.java) for the timer event listener. Therefor it might be possible to extend the already existing [TimerDeclarationImpl](https://github.com/camunda/camunda-bpm-platform/blob/master/engine/src/main/java/org/camunda/bpm/engine/impl/jobexecutor/TimerDeclarationImpl.java).
An instance of the corresponding job declaration should be created inside the `TimerEventListenerItemHandler` when transforming the TimerEventListener. The instance of the job declaration should be set as an property of the activity.

Does it help you for the next step?

Cheers,
Roman

subhrajyotim commented:
Hello Roman,
Not a problem at all, and u are most welcome.
Yes, will do the changes as u mention.

Thanks,
Subhro.

On Wed, Feb 24, 2016 at 3:01 PM, Roman Smirnov <notifications@github.com>
wrote:

> Hi Subhro,
>
> I am very sorry for the late response. I did not receive any notification
> from Github about your answer. Sorry!
>
> I had a look into your pull request. There are some small adjustments
> necessary (see the comments I made). Could you adjust the pull request
> according the comments I made please? Furthermore, please change our coding
> style (see
> https://github.com/camunda/camunda-bpm-platform/tree/master/settings/eclipse
> ).
>
> The next step would be to create a JobDeclaration
> <https://github.com/camunda/camunda-bpm-platform/blob/master/engine/src/main/java/org/camunda/bpm/engine/impl/jobexecutor/JobDeclaration.java>
> for the timer event listener. Therefor it might be possible to extend the
> already existing TimerDeclarationImpl
> <https://github.com/camunda/camunda-bpm-platform/blob/master/engine/src/main/java/org/camunda/bpm/engine/impl/jobexecutor/TimerDeclarationImpl.java>
> .
> An instance of the corresponding job declaration should be created inside
> the TimerEventListenerItemHandler when transforming the
> TimerEventListener. The instance of the job declaration should be set as an
> property of the activity.
>
> Does it help you for the next step?
>
> Cheers,
> Roman
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/camunda/camunda-bpm-platform/pull/201#issuecomment-188161098>
> .
>

subhrajyotim commented:
Hello Roman,
Is there any Idea Intellij 14 Settings similar to
https://github.com/camunda/camunda-bpm-platform/tree/master/settings/eclipse
Please let me know.

Thanks and regards,
Subhro.

On Wed, Feb 24, 2016 at 3:01 PM, Roman Smirnov <notifications@github.com>
wrote:

> Hi Subhro,
>
> I am very sorry for the late response. I did not receive any notification
> from Github about your answer. Sorry!
>
> I had a look into your pull request. There are some small adjustments
> necessary (see the comments I made). Could you adjust the pull request
> according the comments I made please? Furthermore, please change our coding
> style (see
> https://github.com/camunda/camunda-bpm-platform/tree/master/settings/eclipse
> ).
>
> The next step would be to create a JobDeclaration
> <https://github.com/camunda/camunda-bpm-platform/blob/master/engine/src/main/java/org/camunda/bpm/engine/impl/jobexecutor/JobDeclaration.java>
> for the timer event listener. Therefor it might be possible to extend the
> already existing TimerDeclarationImpl
> <https://github.com/camunda/camunda-bpm-platform/blob/master/engine/src/main/java/org/camunda/bpm/engine/impl/jobexecutor/TimerDeclarationImpl.java>
> .
> An instance of the corresponding job declaration should be created inside
> the TimerEventListenerItemHandler when transforming the
> TimerEventListener. The instance of the job declaration should be set as an
> property of the activity.
>
> Does it help you for the next step?
>
> Cheers,
> Roman
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/camunda/camunda-bpm-platform/pull/201#issuecomment-188161098>
> .
>

romansmirnov commented:
Hi Subhro,

Sorry, but we does not have any intellij settings.

The tab and indentation size should be set to 2.

Thanks.

Cheers,
Roman

jangalinski commented:
Hi Subhro,

in some community extensions, we use an .editorconfig file for this, see https://github.com/camunda/camunda-bpm-spring-boot-starter/blob/master/.editorconfig for example.

You will have to install the editorconfig plugin for idea and then everything works fine. 

Regards
Jan

filiphr commented:
Hi,

Just to help a bit. If you use IntelliJ you can use https://plugins.jetbrains.com/plugin/6546 to import the camunda eclipse code formatter. This is what I have used and it works good. Only one important thing to note is that IntelliJ uses different import formatter, so you might need to check that.

Cheers,
Filip

subhrajyotim commented:
Thanks everyone for the pointers and ideas.

.editorconfig mentioned by Jan is what I am using.

It has solved a long standing worry on code conventions for different
projects.

Thanks,
Subhro.

On Fri, Feb 26, 2016 at 1:29 AM, Filip Hrisafov <notifications@github.com>
wrote:

> Hi,
>
> Just to help a bit. If you use IntelliJ you can use
> https://plugins.jetbrains.com/plugin/6546 to import the camunda eclipse
> code formatter. This is what I have used and it works good. Only one
> important thing to note is that IntelliJ uses different import formatter,
> so you might need to check that.
>
> Cheers,
> Filip
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/camunda/camunda-bpm-platform/pull/201#issuecomment-188953380>
> .
>

subhrajyotim commented:
Finally success. Had a little trouble, becoz of changes in JobHandler. Now all ok. Please validate.
Let me know if this is fine and any changes I have to make. 

Thanks again for your time and patience.
Subhro.

romansmirnov commented:
Hi Subhro,

Thanks for updating the pull request. I had a look into it.

I would suggest that I will revise your pull request, so that you can continue on that and implement the next step. In detail I would rework the job declarations, since you have to consider some internals.

I am going to do it either this week or in the beginning of the next week.

Is that okay for you?

Cheers,
Roman

subhrajyotim commented:
Hello Roman,
Thanks a lot for your time. Yes sure, please let me know what I have to do-
no problemo!

".. I would rework the job declarations,"- if there is anything I can help
with please let me know.
I can also try fixing existing bugs, not that I am an expert, but will
definitely give it a shot.

Cheers,
Subhro.


On Mon, Mar 7, 2016 at 3:32 PM, Roman Smirnov <notifications@github.com>
wrote:

> Hi Subhro,
>
> Thanks for updating the pull request. I had a look into it.
>
> I would suggest that I will revise your pull request, so that you can
> continue on that and implement the next step. In detail I would rework the
> job declarations, since you have to consider some internals.
>
> I am going to do it either this week or in the beginning of the next week.
>
> Is that okay for you?
>
> Cheers,
> Roman
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/camunda/camunda-bpm-platform/pull/201#issuecomment-193187943>
> .
>

meyerdan commented:
Hi Subhro,

Just to let you know: Roman is out of office this week (sick). I checked in with him on Friday and he said that he wrote some code on this issue which would allow you to continue the implementation.

He will follow with you next week.

Just as a heads up...

Cheers,
Daniel

subhrajyotim commented:
Thanks Daniel, for letting me know.
Appreciate it!

Thanks and Regards,
Subhro.

On Tue, Mar 15, 2016 at 12:52 AM, Daniel Meyer <notifications@github.com>
wrote:

> Hi Subhro,
>
> Just to let you know: Roman is out of office this week (sick). I checked
> in with him on Friday and he said that he wrote some code on this issue
> which would allow you to continue the implementation.
>
> He will follow with you next week.
>
> Just as a heads up...
>
> Cheers,
> Daniel
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/camunda/camunda-bpm-platform/pull/201#issuecomment-196484643>
> .
>

romansmirnov commented:
Hi Subhro,

Sorry, for my late response.

As promised, I reworked the timer event listener job declaration implementation. The changes can be found on that branch https://github.com/camunda/camunda-bpm-platform/tree/CAM-4450-timer-event-listener and the corresponding commit is https://github.com/camunda/camunda-bpm-platform/commit/91e26d36501e4d5bb63ef5d930c1c62fb6b47589.

I also implemented, that a new timer job is created when the timer event listener is instantiated (see https://github.com/camunda/camunda-bpm-platform/blob/CAM-4450-timer-event-listener/engine/src/main/java/org/camunda/bpm/engine/impl/cmmn/behavior/TimerEventListenerActivityBehavior.java#L28-L29).

In order to complete this feature the following points must be done:
1) Finish the implementation of the timer event listener job declaration:
- Adjust sql scripts: The database table ```ACT_RU_JOB``` now has four new columns (```CASE_EXECUTION_ID_```, ```CASE_INSTANCE_ID_```, ```CASE_DEF_ID_```, ```CASE_DEF_KEY_```). Therefor the engine create sql scripts must be adjusted (see https://github.com/camunda/camunda-bpm-platform/tree/CAM-4450-timer-event-listener/engine/src/main/resources/org/camunda/bpm/engine/db/create)
- Test cases must be written that the timer event listener is initialized correctly.
- Test cases must be written that the job is created and is initialized correctly.
- The properties ```caseExecutionId```, ```caseInstanceId```, ```caseDefinitionId```, ```caseDefinitionKey``` must be exposed in this interface https://github.com/camunda/camunda-bpm-platform/blob/CAM-4450-timer-event-listener/engine/src/main/java/org/camunda/bpm/engine/runtime/Job.java

2) A ```JobDefinition``` must be created, when deploying a case definition containing a timer event listener
3) Implement the timer event listener job that the event ```occurred``` when the timer job is executed.
4) Implement the start trigger
5) Implement repeating timer intervals

I would appreciate it, if you could create for each point a pull request.

Note: If you create a pull request, then please do it against the branch ```CAM-4450-timer-event-listener```.

In the meantime I will close this pull request, since the changes are on the branch ```CAM-4450-timer-event-listener```.

Thanks for your effort!

Cheers,
Roman

subhrajyotim commented:
Got it, Will be starting with #1.

Thanks Roman for your time and attention.

Cheers,
Subhro.

On Mon, Mar 21, 2016 at 3:44 PM, Roman Smirnov <notifications@github.com>
wrote:

> Hi Subhro,
>
> Sorry, for my late response.
>
> As promised, I reworked the timer event listener job declaration
> implementation. The changes can be found on that branch
> https://github.com/camunda/camunda-bpm-platform/tree/CAM-4450-timer-event-listener
> and the corresponding commit is 91e26d3
> <https://github.com/camunda/camunda-bpm-platform/commit/91e26d36501e4d5bb63ef5d930c1c62fb6b47589>
> .
>
> I also implemented, that a new timer job is created when the timer event
> listener is instantiated (see
> https://github.com/camunda/camunda-bpm-platform/blob/CAM-4450-timer-event-listener/engine/src/main/java/org/camunda/bpm/engine/impl/cmmn/behavior/TimerEventListenerActivityBehavior.java#L28-L29
> ).
>
> In order to complete this feature the following points must be done:
> 1) Finish the implementation of the timer event listener job declaration:
>
>    - Adjust sql scripts: The database table ACT_RU_JOB now has four new
>    columns (CASE_EXECUTION_ID_, CASE_INSTANCE_ID_, CASE_DEF_ID_,
>    CASE_DEF_KEY_). Therefor the engine create sql scripts must be
>    adjusted (see
>    https://github.com/camunda/camunda-bpm-platform/tree/CAM-4450-timer-event-listener/engine/src/main/resources/org/camunda/bpm/engine/db/create
>    )
>    - Test cases must be written that the timer event listener is
>    initialized correctly.
>    - Test cases must be written that the job is created and is
>    initialized correctly.
>    - The properties caseExecutionId, caseInstanceId, caseDefinitionId,
>    caseDefinitionKey must be exposed in this interface
>    https://github.com/camunda/camunda-bpm-platform/blob/CAM-4450-timer-event-listener/engine/src/main/java/org/camunda/bpm/engine/runtime/Job.java
>
> 2) A JobDefinition must be created, when deploying a case definition
> containing a timer event listener
> 3) Implement the timer event listener job that the event occurred when
> the timer job is executed.
> 4) Implement the start trigger
> 5) Implement repeating timer intervals
>
> I would appreciate it, if you could create for each point a pull request.
>
> Note: If you create a pull request, then please do it against the branch
> CAM-4450-timer-event-listener.
>
> In the meantime I will close this pull request, since the changes are on
> the branch CAM-4450-timer-event-listener.
>
> Thanks for your effort!
>
> Cheers,
> Roman
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/camunda/camunda-bpm-platform/pull/201#issuecomment-199211079>
>
